### PR TITLE
[WIP] Add the ability to set the dependecies of a mod in mod.conf and add conflicts and provides

### PR DESCRIPTION
**Pull request by TeTpaAka**
TODO:
- [x] move dependencies to mod.conf
- [x] add conflicts to mod.conf
- [x] add provides to mod.conf
- [ ] add preloads to mod.conf
- [x] make the changes visible for the user in the menu
    - [x] load the dependencies also from mod.conf
    - [x] add boxes like the dependencies for the other options
- [x] add description.txt into mod.conf
- [ ] document the changes

implements #3200, though it is still heavily WIP.

### CODE REVIEW

File: src/mods.cpp [hunk 1]
```diff
@@ -61,28 +60,65 @@ void parseModContents(ModSpec &spec)
 
 	// Handle modpacks (defined by containing modpack.txt)
 	std::ifstream modpack_is((spec.path+DIR_DELIM+"modpack.txt").c_str());
-	if(modpack_is.good()){ //a modpack, recursively get the mods in it
+	if (modpack_is.good()) { //a modpack, recursively get the mods in it
```
Comment (rubenwardy) [inline src/mods.cpp]: This confused me, is "is" for "input stream" a common thing? I'd prefer modpack_file.

Comment (ShadowNinja) [inline src/mods.cpp]: `is` is very common.  :-D

**Comment by ShadowNinja** [comment]
Sounds good.  An additional thing to do: move `description.txt` into mod.conf.

File: src/mods.cpp [hunk 2]
```diff
@@ -29,14 +29,13 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "convert_json.h"
 #include "exceptions.h"
 
-static bool parseDependsLine(std::istream &is,
-		std::string &dep, std::set<char> &symbols)
+static bool parseDependsString(std::string &dep,
```
Comment (rubenwardy) [inline src/mods.cpp]: this function looks like it should be called `trimDependsString` for clarity. It trims the string by parsing it.

**Comment by ShadowNinja** [comment]
Oh, you'll also have to add backwards-compatibility by the way -- we can't just break all mods with dependencies suddenly.

File: src/mods.cpp (hunk 2 above)
Comment (rubenwardy) [inline src/mods.cpp]: Oh, it also returns a list of symbols.

**Comment by TeTpaAka** [comment]
It is backwards compatible. If it can't find a field depends in mod.conf it loads it from the depends.txt file like before.

**Comment by 0-afflatus** [comment]
if backwards comatible :+1:

**Comment by sofar** [comment]
It can't be backwards compatible without duplicating depends.txt and description.txt, you will always have all those files present if your mod wants to support both old and new minetest versions.

I'm not saying that that is bad. As long as it's possible to make mods work with older minetest versions, then that's OK with me.

**Comment by LNJ2** [comment]
Still WIP? I would like to use it. It can look very messy, if you split up your code with all the files.

**Comment by TeTpaAka** [comment]
I stopped working on it. There are multiple reasons why:
1. I stopped playing minetest.
2. The logic of preloads is far more complex than I initially thought.

If you only want the existing files to be merged in mod.conf, I could split the pull request, so the working part and be pushed.

**Comment by LNJ2** [comment]
> If you only want the existing files to be merged in mod.conf, I could split the pull request, so the working part and be pushed.

@TeTpaAka That would be nice! :smiley_cat:

**Comment by paramat** [comment]
'conflicts' are mods known to conflict?
What are 'preloads' and how does this differ from 'depends'?
What is 'provides'?
Let's not get fancy here, only the very useful features.

**Comment by TeTpaAka** [comment]
New version: https://github.com/minetest/minetest/pull/4539

**Comment by TeTpaAka** [comment]
Provides would be useful for forks of mods, that have the same abilities and additional functionality, so a mod can either depend on the original, and load, even if the fork is used. Or it can depend directly on the fork.
'preloads' is the opposite of 'depends'. If you want your mod to be loaded before some other.
'conflicts' was just suggested in the issue I tried to address.

Well, anyway. I'm not working on this any longer.

**Pull request closed without merging**